### PR TITLE
Add a 'force' option to MolStandardizer::Uncharger

### PR DESCRIPTION
**Pull request by rvianello**
#### What does this implement/fix? Explain your changes.
I would like to be able to use MolStandardize::Uncharger to neutralize all sites that can be neutralized, also when this resulted in an output structure with an overall non-null charge.

The implemented changes introduce a `force` boolean flag that (when set) disables the check for not-neutralizable positive charges, which would by default result in the attempt to preserve the corresponding negative charges.

#### Any other comments?
I hope the proposed approach makes sense and I didn't overlook any issues.

ah, I removed the copy constructor and the empty destructor in favor of the compiler-generated ones (the copy constructor was also missing the `df_canonicalOrdering` flag). I can of course restore them if preferred.

### CODE REVIEW

File: Code/GraphMol/MolStandardize/Charge.h [hunk 1]
```diff
@@ -130,14 +133,17 @@ class RDKIT_MOLSTANDARDIZE_EXPORT Uncharger {
   Uncharger(bool canonicalOrdering) : Uncharger() {
     df_canonicalOrdering = canonicalOrdering;
   }
-  Uncharger(const Uncharger &other);
```
Comment (greglandrum) [inline Code/GraphMol/MolStandardize/Charge.h:133]: Removing these makes sense, but please add a `default` declaration for them so that it's clear we made an intentional decision

**Comment by greglandrum** [comment]
As written, this is going to do strange and probably undesired things to groups like nitro - `-[N+](=O)[O-]` - or N-oxo - `*-[n+](-[O-])-*`.
I think you probably want to add a check to make sure you aren't removing charges from atoms which are directly bonded to an atom with a matching, but oppositely signed, charge.

**Comment by greglandrum** [comment]
> As written, this is going to do strange and almost certainly undesired things to groups like nitro - `-[N+](=O)[O-]` - or N-oxo - `*-[n+](-[O-])-*`. I think you probably want to add a check to make sure you aren't removing charges from atoms which are directly bonded to an atom with a matching, but oppositely signed, charge.

Ah, never mind. It looks like these are filtered out elsewhere.

**Comment by rvianello** [comment]
yeah, I was a bit worried that I could run into problems with some of those functional groups, but after looking at the code and running some tests I then thought that it could still work as desired. However, I just noticed two other things that don't behave the way I thought, so I'm afraid I'll need to better check these cases first.

**Review by greglandrum — APPROVED** [review]
LGTM

**Comment by greglandrum** [comment]
> yeah, I was a bit worried that I could run into problems with some of those functional groups, but after looking at the code and running some tests I then thought that it could still work as desired. However, I just noticed two other things that don't behave the way I thought, so I'm afraid I'll need to better check these cases first.

Ok, I will hold off on merging until you let me know that you're happy with it.
In the meantime, it would be great if you could add an explicit test to show that something like a nitro group is not actually affected by this change.

File: Code/GraphMol/MolStandardize/Charge.cpp [hunk 2]
```diff
@@ -421,10 +415,9 @@ void Uncharger::unchargeInPlace(RWMol &mol) {
     }
 
     // now do the other negative groups if we still have charges left:
-    neg_surplus = a_matched - q_matched;
     if (neg_surplus > 0) {
       boost::dynamic_bitset<> skipChargeSep(mol.getNumAtoms());
-      for (const auto &pair : n_atoms) {
+      for (const auto &pair : a_atoms) {
```
Comment (rvianello) [inline Code/GraphMol/MolStandardize/Charge.cpp:420]: This one I think was just a bug, maybe a regression. The test on `skipChargeSep` is meant to apply to `a_atoms`, and `n_atoms` is also not a superset of it.

File: Code/GraphMol/MolStandardize/Charge.h (hunk 1 above)
Comment (rvianello) [reply Code/GraphMol/MolStandardize/Charge.h:133]: added.

**Comment by rvianello** [comment]
@greglandrum my initial assumption that the `needsNeutralization` branch could be disabled entirely by the `force` option didn't work with nitrate anions and some other species matched by `neg_acid` and that are not matched by `neg` (but instead contribute to `pos_noh`).

I moved the dependency from the `force` flag inside the `needsNeutralization` branch and I think the issue was this way fixed, together with a bug I found in the process. I added test cases covering the related corner-cases in addition to nitro groups and n-oxides.

**Comment by rvianello** [comment]
@greglandrum in addition to the fixes/adjustments from last Friday, I think I found another small issue in the existing implementation. The expression `n_neg - q_matched` overestimates the number of surplus negative charges because `q_matched` doesn't include the positively charged nitrogen from nitro groups and N-oxides, but the negatively charged oxygens in the same groups contribute to `n_neg`. The side effect is that if an N-oxide is present, the `Uncharger` may incorrectly protonate some other group that is already balanced.

I tried replacing the assignment of `n_neg` with the size of the union of `n_atoms` and `a_atoms` and this fixes the problem when an N-oxide is present, but at the same time it also breaks a couple of test cases where a zwitterionic sulfone `[S+2]([O-])([O-])` is present (afaict it's because in this case the sulphur contributes to `q_matched`, but neither `neg` nor `neg_acid` match the corresponding oxygens).

I'll try and see if there is a reasonably simple solution.

**Comment by rvianello** [comment]
@greglandrum I think this one is now ready for a new review (and hopefully, for merging). I modifed `Uncharger::unchargeInPlace` to perform an upfront merge of the sites from `n_atoms` and `a_atoms`, this way simplifying the related conditional logic and allowing the same execution flow regardless the occurrence of not-neutralizable positive sites and the `force` option (they only differ in whether all or only part of the negative atoms are neutralized). Consequently, I modified `pos_noh` to no longer match the positively charged atom in the nitrate and other anions, and have again a behavior more consistent with `neg` (that doesn't include the negatively charged oxygens from the same functional groups). 

@ptosco fyi, this is the PR I mentioned earlier today.

**Review by greglandrum — APPROVED** [review]
LGTM
sorry it took me so long here

**Comment by rvianello** [comment]
Thank you @greglandrum, sorry it required two reviews :-)
